### PR TITLE
Add two new `abstract` base classes for reading / caching JSON config files

### LINKED ISSUE #23: Create base class to abstract common logic from caching file readers.
See https://github.com/dotnet/docs-actions/pull/21#discussion_r663385277

### PR DESCRIPTION
**Pull request by IEvangelist**
- Prefer instance-based classes over `static` classes
- Additional parameters on methods instead of relying on readers to read every time
- Two new `abstract` base classes:
  - `BaseConfigurationReader`: only requires the `ConfigurationFileName`, reads, and caches parsed config file
  - `BaseMappedConfigurationReader`: exposes custom mapping of the read config file
- General clean up and refactoring

Fixes #23

Attention: @Youssef1313

### CODE REVIEW

File: actions/docs-verifier/BuildVerifier.IO.Abstractions/BuildVerifier.IO.Abstractions.csproj [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
Comment (Youssef1313) [inline actions/docs-verifier/BuildVerifier.IO.Abstractions/BuildVerifier.IO.Abstractions.csproj:4]: Could this be net5.0 and remove `System.Text.Json` reference?

File: actions/docs-verifier/src/ActionRunner/Program.cs [hunk 2]
```diff
@@ -56,15 +70,14 @@ static bool IsRedirectableFile(PullRequestFile file)
...
-    return !isDeletedToc && IsYmlOrMarkdownFile(deletedFileName) && !IsInWhatsNewDirectory(deletedFileName) &&
-        DocfxConfigurationReader.GetMatchers().Any(m => m.Match(deletedFileName).HasMatches);
+    return !isDeletedToc && IsYmlOrMarkdownFile(deletedFileName) && !IsInWhatsNewDirectory(deletedFileName, whatsNewPath) &&
+        matchers.Any(m => m.Match(deletedFileName).HasMatches);
 }
 
 static bool IsYmlOrMarkdownFile(string? fileName) => Path.GetExtension(fileName) is ".yml" or ".md";
 
-static bool IsInWhatsNewDirectory(string? fileName)
+static bool IsInWhatsNewDirectory(string? fileName, string? whatsNewPath)
```
Comment (Youssef1313) [inline actions/docs-verifier/src/ActionRunner/Program.cs:79]: Minor suggestion: I think `fileName` can be made non-nullable, then we won't need null conditional access nor `== true`. For that to be done, I think we'd need to add `[NotNullWhen(true)]` to `fileName` in `IsYmlOrMarkdownFile`.

File: actions/docs-verifier/src/RedirectionVerifier/DocfxConfiguration.cs [hunk 3]
```diff
@@ -7,7 +7,7 @@
 namespace RedirectionVerifier
 {
 #pragma warning disable CA1812 // Avoid uninstantiated internal classes → JSON
```
Comment (Youssef1313) [inline actions/docs-verifier/src/RedirectionVerifier/DocfxConfiguration.cs:9]: Suppression can be removed now.

File: actions/docs-verifier/src/RedirectionVerifier/OpenPublishingRedirections.cs [hunk 4]
```diff
@@ -4,7 +4,7 @@
 namespace RedirectionVerifier
 {
 #pragma warning disable CA1812 // Avoid uninstantiated internal classes → JSON
-    internal sealed class OpenPublishingRedirections
+    public sealed class OpenPublishingRedirections
 #pragma warning restore CA1812 // Avoid uninstantiated internal classes
```
Comment (Youssef1313) [inline actions/docs-verifier/src/RedirectionVerifier/OpenPublishingRedirections.cs:8]: Suppression can be removed.

File: actions/docs-verifier/src/RedirectionVerifier/Redirection.cs [hunk 5]
```diff
@@ -3,13 +3,15 @@
 namespace RedirectionVerifier
 {
 #pragma warning disable CA1812 // Avoid uninstantiated internal classes → JSON
-    internal sealed class Redirection
+    public sealed class Redirection
 #pragma warning restore CA1812 // Avoid uninstantiated internal classes
```
Comment (Youssef1313) [inline actions/docs-verifier/src/RedirectionVerifier/Redirection.cs:7]: Suppression can be removed.

**Review by Youssef1313 — APPROVED** [review]
This is looking very very great!! 🎉

**Review by BillWagner — APPROVED** [review]
This is great work @IEvangelist 

Once you've accepted @Youssef1313 's suggestion, this is ready to :shipit:

File: actions/docs-verifier/src/ActionRunner/Program.cs (hunk 2 above)
Comment (IEvangelist) [reply actions/docs-verifier/src/ActionRunner/Program.cs:79]: Awesome, you're correct - good catch!

File: actions/docs-verifier/BuildVerifier.IO.Abstractions/BuildVerifier.IO.Abstractions.csproj [hunk 6]
```diff
@@ -1,11 +1,8 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.1</TargetFramework>
+    <TargetFramework>net5.0</TargetFramework>
+    <Nullable>enable</Nullable>
```
Comment (Youssef1313) [inline actions/docs-verifier/BuildVerifier.IO.Abstractions/BuildVerifier.IO.Abstractions.csproj:5]: This should be unneeded. Nullable reference types is enabled in Directory.Build.props:

https://github.com/dotnet/docs-actions/blob/c2790ae335d939e53085b4d9f4a1506de9b95078/actions/docs-verifier/Directory.Build.props#L5

```suggestion
```

**Pull request merged**